--- bfd/config.bfd.orig	2020-11-25 16:03:32 UTC
+++ bfd/config.bfd
@@ -1188,6 +1188,11 @@ case "${targ}" in
     want64=true
     ;;
 #ifdef BFD64
+  s390x-*-freebsd*)
+    targ_defvec=s390_elf64_vec
+    targ_selvecs=s390_elf32_vec
+    want64=true
+    ;;
   s390x-*-linux*)
     targ_defvec=s390_elf64_vec
     targ_selvecs=s390_elf32_vec
--- ld/configure.tgt.orig	2020-11-25 16:03:32 UTC
+++ ld/configure.tgt
@@ -776,6 +776,10 @@ rx-*-*)			targ_emul=elf32rx
 s12z-*-*)		targ_emul=m9s12zelf
 			targ_extra_ofiles=ldelfgen.o
 			;;
+s390x-*-freebsd*)	targ_emul=elf64_s390
+			targ_extra_emuls=elf_s390
+			targ_extra_libpath=$targ_extra_emuls
+			tdir_elf_s390=`echo ${targ_alias} | sed -e 's/s390x/s390/'` ;;
 s390x-*-linux*)		targ_emul=elf64_s390
 			targ_extra_emuls=elf_s390
 			targ_extra_libpath=$targ_extra_emuls
