--- gas/configure.orig	2020-09-19 10:19:52 UTC
+++ gas/configure
@@ -7098,6 +7098,7 @@ ia64-*-hpux*)
   ;;
 
 x86_64-*kfreebsd*-gnu|x86_64-*linux*|powerpc*-*linux*| \
+s390*-*freebsd*| \
 s390*-*linux*|s390*-*tpf*|sparc*-*linux*)
   # Find out which ABI we are using.
   echo 'int i;' > conftest.$ac_ext
@@ -7128,6 +7129,9 @@ s390*-*linux*|s390*-*tpf*|sparc*-*linux*)
 	  powerpc64-*linux*)
 	    LD="${LD-ld} -m elf32ppclinux"
 	    ;;
+	  s390x-*freebsd*)
+	    LD="${LD-ld} -m elf_s390"
+	    ;;
 	  s390x-*linux*)
 	    LD="${LD-ld} -m elf_s390"
 	    ;;
@@ -7149,6 +7153,9 @@ s390*-*linux*|s390*-*tpf*|sparc*-*linux*)
 	    ;;
 	  powerpc-*linux*)
 	    LD="${LD-ld} -m elf64ppc"
+	    ;;
+	  s390*-*freebsd*)
+	    LD="${LD-ld} -m elf64_s390"
 	    ;;
 	  s390*-*linux*|s390*-*tpf*)
 	    LD="${LD-ld} -m elf64_s390"
